lib/cn/cn_tree.c

```diff
@@ -1522,50 +1524,69 @@ cn_comp_update_split(
     }
 }
 
+static void
+cn_comp_kvset_append(
+    struct cn_compaction_work *work,
+    struct cn_tree_node       *node,
+    struct kvset              *kvset)
+{
+    struct cn_tree *tree = work->cw_tree;
+
+    if (work->cw_err)
+        return;
+
+    assert(node);
+    assert(kvset);
+
+    rmlock_wlock(&tree->ct_lock);
+
+    assert(node);
+    kvset_list_add(kvset, &node->tn_kvset_list);
+    node->tn_cgen++;
+
+    cn_tree_samp(tree, &work->cw_samp_pre);
+    cn_tree_samp_update_ingest(tree, node);
+    cn_tree_samp(tree, &work->cw_samp_post);
+
+    rmlock_wunlock(&tree->ct_lock);
+}
```

Comment:
Might be better to take the tree lock then process the entire list vs doing them one at a time...

Reply:
Although I suspect in most cases you only have one kvset to add?

Reply:
Right, just one kvset per child. This will change with batched processing.